httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/InetAddressComparator.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.impl.nio;
+
+import org.apache.hc.core5.annotation.Contract;
+import org.apache.hc.core5.annotation.Internal;
+import org.apache.hc.core5.annotation.ThreadingBehavior;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.util.Comparator;
+/**
+ * Comparator that orders IPv6 addresses before IPv4 addresses, and then based on their byte values.
+ */
+@Contract(threading = ThreadingBehavior.STATELESS)
+@Internal
+class InetAddressComparator implements Comparator<InetAddress> {
+
+    /**
+     * Singleton instance of the comparator.
+     */
+    public static final InetAddressComparator INSTANCE = new InetAddressComparator();
+
+    /**
+     * Compares the specified IPv4 and IPv6 addresses.
+     *
+     * @param adr1 the first address to be compared
+     * @param adr2 the second address to be compared
+     * @return a negative integer, zero, or a positive integer as the first argument is less than, equal to, or greater
+     *         than the second
+     * @throws NullPointerException if either argument is null
+     */
+    @Override
+    public int compare(final InetAddress adr1, final InetAddress adr2) {
+        if (adr1 == null && adr2 == null) {
+            return 0;
+        }
+        if (adr1 == null) {
+            return -1;
+        }
+        if (adr2 == null) {
+            return 1;
+        }
+        // Compare IPv6 addresses first
+        if (adr1 instanceof Inet6Address && adr2 instanceof Inet4Address) {
+            return 1;
+        }
+        if (adr1 instanceof Inet4Address && adr2 instanceof Inet6Address) {
+            return -1;
+        }
+        // Compare based on address bytes
```

Comment:
This appears to have nothing to do with the [address sorting algorithm](http://biplane.com.au/blog/?p=141) specified by [RFC 6724](https://www.rfc-editor.org/rfc/rfc6724#section-6), which is [mandated](https://www.rfc-editor.org/rfc/rfc8305#section-4) by RFC 8305. Furthermore, sorting addresses in this way is a bad idea, since it will defeat [round-robin DNS](https://en.wikipedia.org/wiki/Round-robin_DNS), an important load balancing technique.